### DISCUSSION TITLE
Strange behavior of 'film_srf' in specfilm

### DISCUSSION THREAD

**Discussion by UNakade** [mitsuba-renderer/mitsuba3#288] [Q&A] [answered]
Hi. I have been testing out various plugins in Mitsuba 3 (Ubuntu 22.04, `llvm_ad_spectral` variant) and I'm having trouble understanding how a `specfilm` works. Spectra that work in other places don't work as sensor response functions. For example:

`mitsuba.load_dict({'type': 'specfilm', 'ch1': {'type': 'spectrum', 'value': [(590, 1), (595, 1)]}})`

results in 

`RuntimeError: ​[distr_1d.h:477] ContinuousDistribution: needs at least two entries!`

and 

`mitsuba.load_dict({'type': 'specfilm', 'ch1': {'type': 'spectrum', 'value': [(590, 0), (595, 1), (600, 0)]}})`

results in 

`RuntimeError: ​[distr_1d.h:513] ContinuousDistribution: no probability mass found!`

If I use another value at one of the end points or use one extra wavelength,

`mitsuba.load_dict({'type': 'specfilm', 'ch1': {'type': 'spectrum', 'value': [(590, 0), (595, 1), (600, 1)]}})`

 I get:

```
SpecFilm[
  size = [768, 576],
  crop_size = [768, 576],
  crop_offset = [0, 0],
  sample_border = 0,
  filter = GaussianFilter[stddev=0.50, radius=2.00],
  file_format = OpenEXR,
  pixel_format = multichannel,
  component_format = float16,
  film_srf = [
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 2,
        range = [590, 600],
        integral = [5],
        pdf = [0, 1]
      ]
    ]
  ],
  sensor response functions = (
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 3,
        range = [590, 600],
        integral = [7.5],
        pdf = [0, 1, 1]
      ]
    ]
  )
]
```

Here, the size of `film_srf` seems to be one less than that of the sensor response function, but only if it is a regular spectrum:

`mitsuba.load_dict({'type': 'specfilm', 'ch1': {'type': 'spectrum', 'value': [(590, 0), (596, 1), (600, 1)]}})`

results in:

```
SpecFilm[
  size = [768, 576],
  crop_size = [768, 576],
  crop_offset = [0, 0],
  sample_border = 0,
  filter = GaussianFilter[stddev=0.50, radius=2.00],
  file_format = OpenEXR,
  pixel_format = multichannel,
  component_format = float16,
  film_srf = [
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 3,
        range = [590, 600],
        integral = [6.66667],
        pdf = [0, 0.833333, 1]
      ]
    ]
  ],
  sensor response functions = (
    IrregularSpectrum[
      distr = IrregularContinuousDistribution[
        size = 3,
        nodes = [590, 596, 600],
        integral = [7],
        pdf = [0, 1, 1],
      ]
    ]
  )
]
```

I read the documentation. A `specfilm` is supposed to combine all the sensor response functions using inverse transform sampling and then redistribute the samples. Maybe the `film_srf` is that combined distribution. But I still don't understand why it needs to have one less wavelength than the sensor response function.

I know the `specfilm` doesn't inherently have a problem with a sensor response function of the type [(590, 0), (595, 1), (600, 0)], because this worked:

```
mitsuba.load_dict({'type': 'specfilm', 
                   'ch1': {'type': 'spectrum', 'value': [(585, 0), (590, 1), (595, 0)]}, 
                   'ch2': {'type': 'spectrum', 'value': [(590, 0), (595, 1), (600, 0)]}})
```

resulted in:

```
SpecFilm[
  size = [768, 576],
  crop_size = [768, 576],
  crop_offset = [0, 0],
  sample_border = 0,
  filter = GaussianFilter[stddev=0.50, radius=2.00],
  file_format = OpenEXR,
  pixel_format = multichannel,
  component_format = float16,
  film_srf = [
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 3,
        range = [585, 600],
        integral = [7.5],
        pdf = [0, 1, 0]
      ]
    ]
  ],
  sensor response functions = (
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 3,
        range = [585, 595],
        integral = [5],
        pdf = [0, 1, 0]
      ]
    ]
    RegularSpectrum[
      distr = ContinuousDistribution[
        size = 3,
        range = [590, 600],
        integral = [5],
        pdf = [0, 1, 0]
      ]
    ]
  )
]
```

I also noticed that if the spectrum of an area emitter is defined using only two wavelength values, `load_dict` works, but when rendering, an error occurs and Python closes with an error message `Segmentation fault (core dumped)`

**Comment by mcrescas** [comment, marked as answer]
Hi @UNakade ,

I have pushed a fix for the behavior you have encountered. Please let me know if it works for you.
Regarding the area emitter crashing, I was not capable to replicate it. If you have more information, I can take a look also.

**Reply by UNakade** [reply]
Thank you. I could not reproduce the Segmentation fault either. It happens sometimes and in different situations, I don't know yet exactly when
